File: core/storage.py
"""
FMailSender — Storage layer v6.0
All read/write operations for accounts, proxies, recipients, campaign config.
Data lives in %APPDATA%/FMailSender/ (Windows) or ~/FMailSender/ (other).
Works correctly both in dev mode and inside PyInstaller --onefile bundles.
"""
from __future__ import annotations
import dataclasses
import json
import logging
import os
import sys
from pathlib import Path
from cryptography.fernet import Fernet

from core.models import SmtpAccount, Recipient, CampaignConfig

logger = logging.getLogger("fmailsender.storage")


def _get_data_dir() -> Path:
    """Return persistent data directory.

    PyInstaller --onefile: __file__ points to a temporary extraction directory
    that is deleted after the process exits — unsuitable for persistent data.
    Use %APPDATA%/FMailSender on Windows, ~/FMailSender elsewhere.
    """
    if getattr(sys, "frozen", False):
        appdata = os.environ.get("APPDATA", os.path.expanduser("~"))
        return Path(appdata) / "FMailSender"
    return Path(__file__).parent.parent / "data"


DATA_DIR = _get_data_dir()
DATA_DIR.mkdir(parents=True, exist_ok=True)

ACCOUNTS_FILE   = DATA_DIR / "accounts.json"
PROXIES_FILE    = DATA_DIR / "global_proxies.json"
RECIPIENTS_FILE = DATA_DIR / "recipients.json"
CAMPAIGN_FILE   = DATA_DIR / "campaign.json"
KEY_FILE        = DATA_DIR / ".fernet_key"


# ── Encryption ───────────────────────────────────────────────────────────────

def _get_key() -> bytes:
    if KEY_FILE.exists():
        return KEY_FILE.read_bytes()
    key = Fernet.generate_key()
    KEY_FILE.write_bytes(key)
    return key


def _fernet() -> Fernet:
    return Fernet(_get_key())


def _enc(s: str) -> str:
    try:
        return _fernet().encrypt(s.encode()).decode()
    except Exception as exc:
        logger.warning("Encrypt failed: %s", exc)
        return s


def _dec(s: str) -> str:
    try:
        return _fernet().decrypt(s.encode()).decode()
    except Exception as exc:
        logger.warning("Decrypt failed (returning raw): %s", exc)
        return s


# ── Accounts ─────────────────────────────────────────────────────────────────

def save_accounts(accounts: list[SmtpAccount]) -> None:
    data = []
    for a in accounts:
        d = a.to_dict()
        d["proxy"] = ""
        d["proxy_list"] = []
        d["password"] = _enc(d["password"])
        if d.get("access_token"):
            d["access_token"] = _enc(d["access_token"])
        if d.get("refresh_token"):  # FIX SEC-1: refresh_token was stored in plaintext
            d["refresh_token"] = _enc(d["refresh_token"])
        data.append(d)
    ACCOUNTS_FILE.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")


def load_accounts() -> list[SmtpAccount]:
    if not ACCOUNTS_FILE.exists():
        return []
    try:
        data = json.loads(ACCOUNTS_FILE.read_text(encoding="utf-8"))
        accounts = []
        for d in data:
            d["password"] = _dec(d.get("password", ""))
            if d.get("access_token"):
                d["access_token"] = _dec(d["access_token"])
            if d.get("refresh_token"):  # FIX SEC-1: decrypt refresh_token
                d["refresh_token"] = _dec(d["refresh_token"])
            d["proxy"] = ""
            d["proxy_list"] = []
            accounts.append(SmtpAccount.from_dict(d))
        return accounts
    except Exception as exc:
        logger.error("Failed to load accounts from %s: %s", ACCOUNTS_FILE, exc)
        return []


# ── Proxies ──────────────────────────────────────────────────────────────────

_proxy_cache: list[str] = []


def save_proxies(proxies: list[str]) -> None:
    global _proxy_cache
    _proxy_cache = list(proxies)


def load_proxies() -> list[str]:
    return list(_proxy_cache)


# ── Recipients ───────────────────────────────────────────────────────────────

_recipient_cache: list[Recipient] = []


def save_recipients(recipients: list[Recipient]) -> None:
    global _recipient_cache
    _recipient_cache = list(recipients)


def load_recipients() -> list[Recipient]:
    return list(_recipient_cache)


# ── Campaign config ───────────────────────────────────────────────────────────

def save_campaign(cfg: CampaignConfig) -> None:
    CAMPAIGN_FILE.write_text(
        json.dumps(cfg.__dict__, ensure_ascii=False, indent=2), encoding="utf-8"
    )


def load_campaign() -> CampaignConfig:
    if not CAMPAIGN_FILE.exists():
        return CampaignConfig()
    try:
        d = json.loads(CAMPAIGN_FILE.read_text(encoding="utf-8"))
        # FIX COMPAT-1: filter to known fields only — prevents TypeError on
        # version mismatch when campaign.json has extra/removed fields.
        known = {f.name for f in dataclasses.fields(CampaignConfig)}
        d_filtered = {k: v for k, v in d.items() if k in known}
        return CampaignConfig(**d_filtered)
    except Exception as exc:
        logger.error("Failed to load campaign config from %s: %s", CAMPAIGN_FILE, exc)
        return CampaignConfig()

File: backend/src/storage.js
'use strict'
/**
 * FMailSender Storage — Node.js backend
 * AES-256-GCM encryption for passwords/tokens.
 * Data: %APPDATA%/FMailSender/ (Windows) or ~/FMailSender/ (other)
 */
const fs   = require('fs')
const path = require('path')
const crypto = require('crypto')

// ── Data directory ────────────────────────────────────────────────────────────
function getDataDir() {
  if (process.env.FMAIL_DATA_DIR) return process.env.FMAIL_DATA_DIR
  const base = process.env.APPDATA || require('os').homedir()
  return path.join(base, 'FMailSender')
}

const DATA_DIR      = getDataDir()
const KEY_FILE      = path.join(DATA_DIR, '.aes_key')
const ACCOUNTS_FILE = path.join(DATA_DIR, 'accounts.json')
const PROXIES_FILE  = path.join(DATA_DIR, 'global_proxies.json')
const RECIPIENTS_FILE = path.join(DATA_DIR, 'recipients.json')
const CAMPAIGN_FILE = path.join(DATA_DIR, 'campaign.json')
const LICENSE_FILE  = path.join(DATA_DIR, 'license.json')

fs.mkdirSync(DATA_DIR, { recursive: true })

// ── Encryption (AES-256-GCM) ──────────────────────────────────────────────────
let _keyCache = null

function _getKey() {
  if (_keyCache) return _keyCache
  if (fs.existsSync(KEY_FILE)) {
    _keyCache = fs.readFileSync(KEY_FILE)
    return _keyCache
  }
  const key = crypto.randomBytes(32)
  fs.writeFileSync(KEY_FILE, key, { mode: 0o600 })
  _keyCache = key
  return key
}

function encrypt(plaintext) {
  if (!plaintext) return plaintext
  try {
    const key = _getKey()
    const iv  = crypto.randomBytes(12)
    const cipher = crypto.createCipheriv('aes-256-gcm', key, iv)
    const enc  = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()])
    const tag  = cipher.getAuthTag()
    return Buffer.concat([iv, tag, enc]).toString('base64')
  } catch (e) {
    return plaintext
  }
}

function decrypt(ciphertext) {
  if (!ciphertext) return ciphertext
  try {
    const buf  = Buffer.from(ciphertext, 'base64')
    const key  = _getKey()
    const iv   = buf.slice(0, 12)
    const tag  = buf.slice(12, 28)
    const enc  = buf.slice(28)
    const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv)
    decipher.setAuthTag(tag)
    return Buffer.concat([decipher.update(enc), decipher.final()]).toString('utf8')
  } catch (e) {
    // possibly plain text (legacy / migration)
    return ciphertext
  }
}

// ── Safe JSON read/write ──────────────────────────────────────────────────────
function readJson(file, fallback) {
  try {
    if (!fs.existsSync(file)) return fallback
    return JSON.parse(fs.readFileSync(file, 'utf8'))
  } catch { return fallback }
}

function writeJson(file, data) {
  const tmp = file + '.tmp'
  fs.writeFileSync(tmp, JSON.stringify(data, null, 2), 'utf8')
  fs.renameSync(tmp, file)
}

// ── Accounts ──────────────────────────────────────────────────────────────────
function saveAccounts(accounts) {
  const data = accounts.map(a => {
    const d = { ...a }
    if (d.password)      d.password      = encrypt(d.password)
    if (d.access_token)  d.access_token  = encrypt(d.access_token)
    if (d.refresh_token) d.refresh_token = encrypt(d.refresh_token)
    return d
  })
  writeJson(ACCOUNTS_FILE, data)
}

function loadAccounts() {
  const data = readJson(ACCOUNTS_FILE, [])
  return data.map(d => {
    const a = { ...d }
    a.password      = decrypt(a.password || '')
    if (a.access_token)  a.access_token  = decrypt(a.access_token)
    if (a.refresh_token) a.refresh_token = decrypt(a.refresh_token)
    return a
  })
}

// ── Proxies ───────────────────────────────────────────────────────────────────
let _proxyCache = null

function saveProxies(proxies) {
  _proxyCache = [...proxies]
  writeJson(PROXIES_FILE, proxies)
}

function loadProxies() {
  if (_proxyCache) return [..._proxyCache]
  const data = readJson(PROXIES_FILE, [])
  _proxyCache = data.filter(Boolean)
  return [..._proxyCache]
}

// ── Recipients ────────────────────────────────────────────────────────────────
function saveRecipients(recipients) { writeJson(RECIPIENTS_FILE, recipients) }
function loadRecipients()           { return readJson(RECIPIENTS_FILE, []) }

// ── Campaign config ───────────────────────────────────────────────────────────
const CAMPAIGN_DEFAULTS = {
  subject: '', body_html: '', body_text: '',
  from_name: '', reply_to: '',
  delay_min: 1.0, delay_max: 3.0,
  daily_limit_per_account: 500
}

function saveCampaign(cfg) { writeJson(CAMPAIGN_FILE, cfg) }
function loadCampaign()    { return { ...CAMPAIGN_DEFAULTS, ...readJson(CAMPAIGN_FILE, {}) } }

// ── License ───────────────────────────────────────────────────────────────────
function saveLicenseCache(data) { writeJson(LICENSE_FILE, data) }
function loadLicenseCache()     { return readJson(LICENSE_FILE, null) }

module.exports = {
  DATA_DIR, LICENSE_FILE,
  encrypt, decrypt,
  saveAccounts, loadAccounts,
  saveProxies,  loadProxies,
  saveRecipients, loadRecipients,
  saveCampaign,   loadCampaign,
  saveLicenseCache, loadLicenseCache,
}
